File: Tests.Common.VB/CompilerGenerated/Member.vb
﻿Partial Public Class VBCompilerGeneratedBase
    <Fact> <Trait("Category", Member)>
    Sub InstanceMember()
        Dim s = ""
        RunTest(
            Function() s.Length,
            "() => s.Length",
            "Function() s.Length"
        )
    End Sub

    <Fact> <Trait("Category", Member)>
    Sub ClosedVariable()
        Dim s = ""
        RunTest(
            Function() s,
            "() => s",
            "Function() s"
        )
    End Sub

    <Fact> <Trait("Category", Member)>
    Sub StaticMember()
        RunTest(
            Function() String.Empty,
            "() => string.Empty",
            "Function() String.Empty"
        )
    End Sub

    <Fact(Skip:="Test for nested scope")> <Trait("Category", Member)>
    Sub NestedClosureScope()
        Assert.False(True)
    End Sub
End Class

File: Tests.Common/CompilerGenerated/Member.cs
﻿using Xunit;
using static ExpressionToString.Tests.Categories;

namespace ExpressionToString.Tests {
    public partial class CompilerGeneratedBase {
        [Fact]
        [Trait("Category", Member)]
        public void InstanceMember() {
            var s = "";
            RunTest(
                () => s.Length,
                "() => s.Length",
                "Function() s.Length", 
                @"Lambda(
    MakeMemberAccess(s,
        typeof(string).GetProperty(""Length"")
    )
)"
            );
        }

        [Fact]
        [Trait("Category", Member)]
        public void ClosedVariable() {
            var s = "";
            RunTest(
                () => s,
                "() => s",
                "Function() s", 
                "Lambda(s)"
            );
        }

        [Fact]
        [Trait("Category", Member)]
        public void StaticMember() => RunTest(
            () => string.Empty,
            "() => string.Empty",
            "Function() String.Empty", 
            @"Lambda(
    MakeMemberAccess(null,
        typeof(string).GetField(""Empty"")
    )
)"
        );

        [Fact(Skip ="Test for nested closure scopes")]
        [Trait("Category", Member)]
        public void NestedClosedVariable() {
            Assert.False(true);
        }
    }
}
